libtenzir/builtins/operators/ocsf.cpp

```diff
@@ -0,0 +1,228 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2025 The Tenzir Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include "tenzir/ocsf.hpp"
+
+#include "tenzir/concept/printable/tenzir/json.hpp"
+#include "tenzir/modules.hpp"
+#include "tenzir/tql2/plugin.hpp"
+#include "tenzir/view3.hpp"
+
+namespace tenzir::plugins::ocsf {
+namespace {
+
+auto cast(const arrow::Array& array, const type& ty)
+  -> std::shared_ptr<arrow::Array>;
+
+template <basic_type Type>
+auto cast(const type_to_arrow_array_t<Type>& array, const Type&)
+  -> std::shared_ptr<type_to_arrow_array_t<Type>> {
+  return std::make_shared<type_to_arrow_array_t<Type>>(array.data());
+}
+
+auto cast(const enumeration_type::array_type&, const enumeration_type&)
+  -> std::shared_ptr<enumeration_type::array_type> {
+  TENZIR_UNREACHABLE();
+}
+
+auto cast(const arrow::MapArray&, const map_type&)
+  -> std::shared_ptr<arrow::MapArray> {
+  TENZIR_UNREACHABLE();
+}
+
+auto cast(const arrow::ListArray& array, const list_type& ty)
+  -> std::shared_ptr<arrow::ListArray> {
+  auto values = cast(*array.values(), ty.value_type());
+  // TODO: What about `array.offset()`?
+  return check(arrow::ListArray::FromArrays(
+    *array.offsets(), *values, arrow::default_memory_pool(),
+    array.null_bitmap(), array.data()->null_count));
+}
+
+auto cast(const arrow::StructArray& array, const record_type& ty)
+  -> std::shared_ptr<arrow::StructArray> {
+  auto fields = arrow::FieldVector{};
+  auto field_arrays = arrow::ArrayVector{};
+  for (auto&& field : ty.fields()) {
+    fields.push_back(field.type.to_arrow_field(field.name));
+    auto field_array = array.GetFieldByName(std::string{field.name});
+    if (not field_array) {
+      field_arrays.push_back(check(
+        arrow::MakeArrayOfNull(field.type.to_arrow_type(), array.length())));
+      continue;
+    }
+    field_arrays.push_back(cast(*field_array, field.type));
+  }
+  return make_struct_array(array.length(), array.null_bitmap(), fields,
+                           field_arrays);
+}
+
+auto cast(const arrow::Array& array, const type& ty)
+  -> std::shared_ptr<arrow::Array> {
+  if (is<string_type>(ty) and ty.attribute("print_json")) {
+    if (is<arrow::StringArray>(array)) {
+      // Keep strings as they are (assuming they are already JSON).
+      return std::make_shared<arrow::StringArray>(array.data());
+    }
+    // Convert everything else into JSON.
+    auto builder = arrow::StringBuilder{};
+    auto printer = json_printer{{.style = no_style(), .oneline = true}};
```

Comment:
Your PR description says `print_json()` but it `.online = true` suggests NDJSON. JMS that this is intended. I think NDJSON is the better choice by far.

Reply:
Oh yeah. I find it a bit weird that `.print_json()` does indentation and everything, but I guess it's consistent… PR description fixed, actual behavior should of course match `.print_ndjson()` (seems like you would almost always want to use that instead of the non-nd version…).